MIDAS/src/hardware/telemetry.cpp

```diff
@@ -0,0 +1,425 @@
+    }
+    Serial.println("[DEBUG]: Radio Initialized");
+
+    // Defaults after init are 434.0MHz, 13dBm, Bw = 125 kHz, Cr = 4/5, Sf =
+    // 128chips/symbol, CRC on
+
+    if (!rf95.setFrequency(RF95_FREQ)) {
+        return ErrorCode::RADIO_SET_FREQUENCY_FAILED;
+    }
+
+    /*
+     * The default transmitter power is 13dBm, using PA_BOOST.
+     * If you are using RFM95/96/97/98 modules which uses the PA_BOOST
+     * transmitter pin, then you can set transmitter powers from 5 to 23 dBm:
+     */
+    rf95.setTxPower(6, false);
+
+    sei();
+#endif
+    return ErrorCode::NoError;
+}
+
+#if defined(ENABLE_TELEMETRY) && !defined(ENABLE_SILSIM_MODE)
+Telemetry::Telemetry() : rf95(RFM96_CS, RFM96_INT, hardware_spi1) {}
+#else
+Telemetry::Telemetry() {}
+#endif
+
+/**
+ * @brief  This function handles commands sent from the ground station
+ * to TARS. The effects of this function depend on the command
+ * sent.
+ *
+ * @param cmd: struct containing information necessary to process
+ *             ground station command.
+ *
+ * @return void
+ */
+void Telemetry::handleCommand(const telemetry_command &cmd) {
+    /* Check if the security code is present and matches on ground and on the
+     * rocket */
+    if (cmd.verify != std::array<char, 6>{'A', 'Y', 'B', 'E', 'R', 'K'}) {
+        return;
+    }
+    /* Check if lasted command ID matched current command ID */
+    if (last_command_id == cmd.cmd_id) {
+        return;
+    }
+    last_command_id = (int16_t)cmd.cmd_id;
+
+    /*
+     * Write frequency to SD card to save
+     * between runs
+     */
+    if (cmd.command == SET_FREQ) {
+        freq_status.should_change = true;
+        freq_status.new_freq = cmd.freq;
+    }
+
+    if (cmd.command == SET_CALLSIGN) {
+        memcpy(callsign, cmd.callsign, sizeof(cmd.callsign));
+        Serial.println("[DEBUG]: Got callsign");
+    }
+
+    if (cmd.command == ABORT) {
+        if (!abort) {
+            abort = !abort;
+        }
+        Serial.println("[DEBUG]: Got abort");
+    }
+}
+
+// #define TLM_DEBUG
+
+/**
+ * @brief This function transmits data from the struct provided as
+ * the parameter (data collected from sensor suite) to the
+ * ground station. The function also switches to a new commanded
+ * frequency based on a previously received command and waits for
+ * a response from the ground station.
+ *
+ * @param sensor_data: struct of data from the sensor suite to be
+ *                     transmitted to the ground station.
+ *
+ * @return void
+ */
+void Telemetry::transmit() {
+#ifdef ENABLE_TELEMETRY
+#ifdef TLM_DEBUG
+    const uint8_t data[4] = {0, 1, 2, 3};
+    rf95.send(data, 4);
+    Serial.println("Sending packet...");
+    rf95.waitPacketSent();
+    Serial.println("Sent packet");
+#else
+    static bool blue_state = false;
+    digitalWrite(LED_BLUE, blue_state);
+    blue_state = !blue_state;
+    
+
+    TelemetryPacket packet = makePacket(); //change to new data struct
```

Comment:
makePacket() has an argument but none is passed in. You'll probably have to modify transmit to take in the sensor data struct and pass that into make packet as well